### PR TITLE
Code modify to make train_gnmt support multi gpus training

### PR DESCRIPTION
**Pull request by pengxin99**
## Description ##
Modify the train_gnmt.py to adapt the multi gpus environment. The change is according to the train_transformer.py which can training with multi gpus.
### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szha** [comment]
@pengxin99 this is awesome. Thanks for the contribution! Looks like git wasn't able to do an auto-merge when checking out. Could you try the following for a rebase?
```bash
git remote add dmlc https://github.com/dmlc/gluon-nlp
pull dmlc master --rebase
git push --force
```

**Comment by pengxin99** [comment]
@szha Thanks, and i will make appropriate modify.

File: scripts/machine_translation/train_gnmt.py [hunk 1]
```diff
@@ -195,33 +202,46 @@ def train():
...
-            tgt_seq = tgt_seq.as_in_context(ctx)
-            src_valid_length = src_valid_length.as_in_context(ctx)
-            tgt_valid_length = tgt_valid_length.as_in_context(ctx)
+
+            src_wc, tgt_wc, bs = np.sum([(shard[2].sum(), shard[3].sum(), shard[0].shape[0])
+                                         for shard in seqs], axis=0)
+
+            seqs = [[seq.as_in_context(_context) for seq in shard]
+                    for _context, shard in zip(context, seqs)]
+            src_wc = src_wc.asscalar()
```
[the remark this answers is not in the document]
Comment (pengxin99) [reply scripts/machine_translation/train_gnmt.py]: thanks, and I will remove the redundant

File: scripts/machine_translation/train_gnmt.py [hunk 2]
```diff
@@ -59,47 +59,51 @@
...
-parser.add_argument('--epochs', type=int, default=40, help='upper epoch limit')
-parser.add_argument('--num_hidden', type=int, default=128, help='Dimension of the embedding '
+parser.add_argument('--dataset', type=str, default='IWSLT2015', help='Dataset to use, default is IWSLT2015.')
+parser.add_argument('--src_lang', type=str, default='en', help='Source language, default is en.')
+parser.add_argument('--tgt_lang', type=str, default='vi', help='Target language, default is vi.')
+parser.add_argument('--epochs', type=int, default=40, help='upper epoch limit, , default is 40.')
+parser.add_argument('--num_hidden', type=int, default=128, help='Dimension of the rnn hidden states. default is 128.')
+parser.add_argument('--num_embedding', type=int, default=128, help='Dimension of the embedding, default is 128. '
                                                                 'vectors and states.')
+parser.add_argument('--attention_type', type=str, default='dot', help=' Attention type, default is dot.')
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: Default one is scaled_luong.

File: scripts/machine_translation/train_gnmt.py [hunk 3]
```diff
@@ -195,33 +202,44 @@ def train():
...
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+                    loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
+                    loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
+                    LS.append(loss)
+
+            for L in LS:
+                L.backward()
+
             grads = [p.grad(ctx) for p in model.collect_params().values()]
             gnorm = gluon.utils.clip_global_norm(grads, args.clip)
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: Currently, this function does not support clipping in multi-gpu context.

File: scripts/machine_translation/train_gnmt.py [hunk 4]
```diff
@@ -195,33 +202,44 @@ def train():
...
+
             with mx.autograd.record():
-                out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
-                loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
-                loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
-                loss.backward()
+                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+                    loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
+                    loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: The loss need to be carefully rescaled, so that the final gradient is the average of the gradients w.r.t. tokens.

File: scripts/machine_translation/train_gnmt.py [hunk 5]
```diff
@@ -195,33 +202,44 @@ def train():
...
             gnorm = gluon.utils.clip_global_norm(grads, args.clip)
             trainer.step(1)
-            src_wc = src_valid_length.sum().asscalar()
-            tgt_wc = (tgt_valid_length - 1).sum().asscalar()
-            step_loss = loss.asscalar()
-            log_avg_loss += step_loss
+
+            # calculate the loss at all contexts
+            step_loss = sum([L.asscalar() for L in LS])
+            log_avg_loss += step_loss / len(context)
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: Similar to the above comment, the scaling factor needs to be calibrated.

**Comment by pengxin99** [comment]
@szhengac Thanks for review,
>  The loss need to be carefully rescaled, so that the final gradient is the average of the gradients w.r.t. tokens. 

and i want to know does this mean every loss at each context should be average at global batch?

**Comment by szhengac** [comment]
@pengxin99 Yes, the gradient needs to be averaged by the total number of tokens across all the GPUs.

File: scripts/machine_translation/train_gnmt.py [hunk 6]
```diff
@@ -195,48 +217,58 @@ def train():
...
+                    for _context, shard in zip(context, seqs)]
+
+            LS = []
             with mx.autograd.record():
-                out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
-                loss = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).mean()
-                loss = loss * (tgt_seq.shape[1] - 1) / (tgt_valid_length - 1).mean()
-                loss.backward()
-            grads = [p.grad(ctx) for p in model.collect_params().values()]
-            gnorm = gluon.utils.clip_global_norm(grads, args.clip)
```
Comment (sxjscience) [inline scripts/machine_translation/train_gnmt.py]: Why remove the norm clipping?

File: scripts/machine_translation/train_gnmt.py [hunk 7]
```diff
@@ -195,48 +217,58 @@ def train():
...
-            log_avg_gnorm += gnorm
+                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+
+                    ls = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).sum()
+                    LS.append(ls * (tgt_seq.shape[1] - 1))
+            for L in LS:
+                L.backward()
+
+            trainer.step(float(loss_denom))
```
Comment (sxjscience) [inline scripts/machine_translation/train_gnmt.py]: We need to divide `loss_denom()` before calling `step()`, i.e, always divide first and call `step(1.0)`. This ensures that the head_gradient will not be too large. If the head gradient have a large scale, the calculated gradient will also have a large scale and may explode in the backpropagation process.

Comment (pengxin99) [reply scripts/machine_translation/train_gnmt.py]: Thanks for review, I see the transformer do the step() with `loss_denom()`, and i think step() function will make loss averaged by the input parameter(loss_denorm), so the gradient will not be too large, am i right?

- transformer code

https://github.com/dmlc/gluon-nlp/blob/1a5f9f38c9f2d070a8f6af6a8234d8e1daed5778/scripts/machine_translation/train_transformer.py#L321

- source code of step()

https://github.com/apache/incubator-mxnet/blob/d2102faa228bdc6723a9da299c6ff5999cbbdcdb/python/mxnet/gluon/trainer.py#L276

**Comment by mli** [comment]
Job PR-435/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-435/9/index.html

File: scripts/machine_translation/train_gnmt.py (hunk 7 above)
Comment (szhengac) [reply scripts/machine_translation/train_gnmt.py]: If the loss is too large, the step `loss.backward()` can explode and lead to NAN. That's why we rescaled the loss.

**Comment by pengxin99** [comment]
@szhengac @sxjscience could you please review the lasted commit if these changes work well, thanks~

**Comment by sxjscience** [comment]
Looks good. I'll approve after I confirmed that the results are the same.

File: scripts/machine_translation/train_gnmt.py [hunk 8]
```diff
@@ -195,38 +217,56 @@ def train():
...
-            grads = [p.grad(ctx) for p in model.collect_params().values()]
+                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
+                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)
+
+                    ls = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).sum()
+                    LS.append(ls * (tgt_seq.shape[1] - 1) / float(loss_denom))
+            for L in LS:
+                L.backward()
+
+            grads = [p.grad(c) for p in model.collect_params().values() for c in context]
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: The way you do is computing norm: (a^2 + b^2)/c^2, where a and b is the grad in context 0 and 1, respectively, and c is the scaling factor. However, it is supposed to be (a + b)^2/c^2.

Comment (szhengac) [reply scripts/machine_translation/train_gnmt.py]: You can check this by taking a look at clip_global_norm function.

Comment (szhengac) [reply scripts/machine_translation/train_gnmt.py]: So my suggestion is to write another clip_global_norm function, where we can take as input a 2-D array, and we first sum up each row and then compute the global norm.

Comment (szha) [reply scripts/machine_translation/train_gnmt.py]: Let's write a parameter version of clip_global_norm, so that the norms are correctly aggregated by parameters.

File: scripts/machine_translation/train_gnmt.py [hunk 9]
```diff
@@ -190,43 +214,94 @@ def evaluate(data_loader):
...
+
+    assert len(arrays) > 0
+    assert context_len >= 1
+
+    ctx = arrays[0].context
+    arrays_at_all_context = arrays[0]
+    # multi GPUs need to sum up each row(parameters at each context) first
+    if context_len > 1:
+        for i in range(context_len):
+            arrays_at_all_context += arrays[i]
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: What is this here? Sum up the first context_len elements?

File: scripts/machine_translation/train_gnmt.py [hunk 10]
```diff
@@ -190,43 +214,94 @@ def evaluate(data_loader):
...
+        if array.stype == 'default':
+            x = array.reshape((-1,))
+            return ndarray.dot(x, x)
+        return array.norm().square()
+
+    assert len(arrays) > 0
+    assert context_len >= 1
+
+    ctx = arrays[0].context
+    arrays_at_all_context = arrays[0]
```
Comment (szhengac) [inline scripts/machine_translation/train_gnmt.py]: As you use input `grads = [p.grad(c) for p in model.collect_params().values() for c in context]` in line 290, this is the grad of some parameter at the first context. In line 236, you sum up the gradients of this parameter at all contexts, and the gradient at the first context is doubled.

File: scripts/machine_translation/train_gnmt.py (hunk 9 above)
Comment (pengxin99) [reply scripts/machine_translation/train_gnmt.py]: This code is mean to sum up the same parameter at different contexts. For these parameters is Ndarray type, so i use `+=` derictly

File: scripts/machine_translation/train_gnmt.py (hunk 10 above)
Comment (szhengac) [reply scripts/machine_translation/train_gnmt.py]: The grads is 1-D list. So you have 'arrays_at_all_context == grads[0] '.

**Comment by mli** [comment]
Job PR-435/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-435/13/index.html

**Comment by pengxin99** [comment]
@szha could you please help me found what is the problem or where did i go wrong?  Thanks~
I use the new clip gradient ([pr #470](https://github.com/dmlc/gluon-nlp/pull/470)) for this multi GPUs training, but i got this error :

> Parameter 'gnmt_enc_rnn0_l_i2h_weight' was not initialized on context cpu(0). It was only initialized on [gpu(0), gpu(1), gpu(2), gpu(3)]

i think it maybe the same as Mxnet issue [Cannot call data() on parameters initialized on multiple GPUs](https://github.com/apache/incubator-mxnet/issues/11329)

**main code** 

- parameter init code

```
context = [mx.cpu()] if args.gpus is None or args.gpus == '' else \
    [mx.gpu(int(x)) for x in args.gpus.split(',')]
ctx = context[0]

# load GNMT model
encoder, decoder = get_gnmt_encoder_decoder(hidden_size=args.num_hidden,
                                            dropout=args.dropout,
                                            num_layers=args.num_layers,
                                            num_bi_layers=args.num_bi_layers,
                                            attention_cell=args.attention_type)

model = NMTModel(src_vocab=src_vocab, tgt_vocab=tgt_vocab, encoder=encoder, decoder=decoder,
                 embed_size=args.num_embedding, prefix='gnmt_')


model.initialize(init=mx.init.Uniform(0.1), ctx=context)
static_alloc = True
model.hybridize(static_alloc=static_alloc)
```

- call clip_grad_global_norm code
```
            with mx.autograd.record():
                for src_seq, tgt_seq, src_valid_length, tgt_valid_length in seqs:
                    out, _ = model(src_seq, tgt_seq[:, :-1], src_valid_length, tgt_valid_length - 1)

                    ls = loss_function(out, tgt_seq[:, 1:], tgt_valid_length - 1).sum()
                    LS.append(ls * (tgt_seq.shape[1] - 1) / float(loss_denom))
            for L in LS:
                L.backward()

            # grads = [p.grad(c) for p in model.collect_params().values() for c in context]
            # gnorm = clip_global_norm(grads, args.clip, context_len=len(context))
            trainer.allreduce_grads()
            gnorm = clip_grad_global_norm(model.collect_params().values(), args.clip)
            trainer.update(args.batch_size)

            log_avg_gnorm += gnorm
```

**Comment by szha** [comment]
@pengxin99 the problem was that the parameter's get method is relying on current_context when calling grad() without argument. I will post a fix in GluonNLP first and see how to resolve it upstream.

**Comment by pengxin99** [comment]
@szha Thanks, and I will test GNMT performancem with multi GPUS , and then update this pr as soon as possible.

**Comment by mli** [comment]
Job PR-435/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-435/14/index.html

**Comment by mli** [comment]
Job PR-435/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-435/15/index.html

**Comment by mli** [comment]
Job PR-435/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-435/16/index.html

**Comment by szha** [comment]
@eric-haibin-lin will help continue the work here. cc @pengxin99
